tools/ClassGen/NodeBuilder.cpp

```diff
@@ -675,25 +793,59 @@ void NodeBuilder::emitExportMethods(std::ostream &os) const {
   os << "  opProto->set_op_type(\"Glow_" << name_ << "\");\n";
   os << "  opProto->set_name(glow::legalizeName(N__->getName()));\n";
 
-  // Add all of the node's inputs.
-  for (const auto &op : nodeInputs_) {
-    os << "  opProto->add_input(N__->get" << op
-       << "().generateNodeOutputName(/* stripResNoFor0thInput */ true));\n";
-    // Note: Add each input's type attributes so that other tools have easy
-    // visibility into types. This info may go ignored by the importer.
-    os << "  addTypeAttributes(opProto, N__, " << name_ << "Node::" << op
-       << "Idx, /* isInput */ true);\n";
-  }
+  // In order to manage variable inputs and outputs and export a clean ONNX
+  // model, the export is done with a generic code working on the Node object,
+  // the variability being well managed in the node getter function.
+  os << "  for(unsigned i=0; i<node->getNumInputs(); i++) {\n";
+  os << "    opProto->add_input(node->getNthInput(i).generateNodeOutputName(/* "
+        "stripResNoFor0thInput */ true));\n";
+  os << "    addTypeAttributes(opProto, N__, i, /* isInput */ true);\n";
+  os << "  }\n";
+  os << "  for(unsigned i=0; i<node->getNumResults(); i++) {\n";
+  os << "    "
+        "opProto->add_output(node->getNthResult(i).generateNodeOutputName(/* "
+        "stripResNoFor0thInput */ true));\n";
+  os << "    addTypeAttributes(opProto, N__, i, /* isInput */ false);\n";
+  os << "  }\n";
 
-  // Add all of the node's outputs.
-  for (const auto &op : nodeOutputs_) {
-    os << "  opProto->add_output(N__->get" << op.second
-       << "().generateNodeOutputName(/* stripResNoFor0thInput */ true));\n";
-    // Note: export the type attributes even if not needed by the importer, so
-    // that other tools have easy visibility into types. This info may go
-    // ignored by the importer.
-    os << "  addTypeAttributes(opProto, N__, " << name_ << "Node::" << op.second
-       << "Idx, /* isInput */ false);\n";
+  // Variadic parameters are always the last params. But in case there are
+  // multiple variadic inputs or outputs, there is the need to know their index
+  // and size in the input or output list. This information is exported as
+  // extra attributes.
+  bool hasVariadicInput = false;
+  for (const auto &op : members_) {
+    if ((op.first).type != MemberType::VectorNodeValue) {
+      continue;
+    }
+    if (!hasVariadicInput) {
+      os << "  // Handle variadic inputs\n";
+      os << "  {\n";
+      os << "    unsigned_t index = " << nodeInputs_.size() << ";\n";
+      hasVariadicInput = true;
+    }
+    std::string sParamSize = "N__->get" + op.second + "().size()";
+    os << "    addValueAttribute(opProto, \"" << op.second
+       << "_Index\", index);\n";
+    os << "    addValueAttribute(opProto, \"" << op.second << "_Size\", "
+       << sParamSize << ");\n";
+    os << "    index += " + sParamSize + ";\n";
+  }
+  if (hasVariadicInput) {
+    os << "  }\n";
+  }
```

Comment:
Perhaps I'm missing something -- why do we need to export this extra variadic inputs info here w/ Index and Size? We already handle variadic inputs when we write out node members (e.g. for nodes like concat) here, right?

https://github.com/pytorch/glow/blob/61da45dfd1a7a563b68de5584c4a2237b1ce9d47/tools/ClassGen/NodeBuilder.cpp#L699-L712